client/json-rpc/rpc_procedures.c

```diff
@@ -0,0 +1,328 @@
+	char address[128] = {0};
+	if (params) {
+		if (cJSON_IsArray(params)) {
+			size_t size = cJSON_GetArraySize(params);
+			int i = 0;
+			for (i = 0; i < size; i++) {
+				cJSON* item = cJSON_GetArrayItem(params, i);
+				if(cJSON_IsString(item)) {
+					strcpy(address, item->valuestring);
+					break;
+				}
+			}
+		} else {
+			ctx->error_code = 1;
+			ctx->error_message = strdup("Invalid parameters.");
+			return NULL;
+		}
+	}
+	
+	if(g_xdag_state < XDAG_STATE_XFER) {
+		ctx->error_code = 1;
+		ctx->error_message = strdup("Not ready to show a balance.");
+		return NULL;
+	} else {
+		cJSON * ret = NULL;
+		cJSON* item = cJSON_CreateObject();
+		xdag_hash_t hash;
+		xdag_amount_t balance;
+		
+		if(strlen(address)) {
+			xdag_address2hash(address, hash);
+			balance = xdag_get_balance(hash);
+		} else {
+			balance = xdag_get_balance(0);
+		}
+		
+		char str[128] = {0};
+		sprintf(str, "%.9Lf",  xdag_amount2xdag(balance) + (long double)xdag_amount2cheato(balance) / 1000000000);
+		cJSON_AddItemToObject(item, "balance", cJSON_CreateString(str));
+		
+		ret = cJSON_CreateArray();
+		cJSON_AddItemToArray(ret, item);
+		return ret;
+	}
+	
+	return NULL;
+}
+
+
+/* xfer */
+/*
+ request:
+ "method":"do_xfer", "params":[{"amount":"1.0", "address":"AAAAAAAAAAAAAAAAA"}], "id":1
+ "jsonrpc":"2.0", "method":"do_xfer", "params":["S","A","sign"], "id":1
+ "version":"1.1", "method":"do_xfer", "params":["S","A","sign"], "id":1
+ 
+ reponse:
+ "result":[{"block":"AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"}], "error":null, "id":1
+ "jsonrpc":"2.0", "result":[{"block":"AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"}], "error":null, "id":1
+ "version":"1.1", "result":[{"block":"AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"}], "error":null, "id":1
+ */
+
+cJSON * method_xdag_do_xfer(struct xdag_rpc_context * ctx, cJSON * params, cJSON *id, char *version);
+cJSON * method_xdag_do_xfer(struct xdag_rpc_context * ctx, cJSON * params, cJSON *id, char *version)
+{
+	//todo: need password or not?
+	xdag_debug("rpc call method do_xfer, version %s", version);
+	
+	char amount[128] = {0};
+	char address[128] = {0};
+	
+	if (params) {
+		if (cJSON_IsArray(params) && cJSON_GetArraySize(params) == 1) {
+			cJSON* param = cJSON_GetArrayItem(params, 0);
+			if (!param || !cJSON_IsObject(param)) {
+				ctx->error_code = 1;
+				ctx->error_message = strdup("Invalid parameters.");
+				return NULL;
+			}
+			
+			cJSON * json_amount = cJSON_GetObjectItem(param, "amount");
+			if (cJSON_IsString(json_amount)) {
+				strcpy(amount, json_amount->valuestring);
+			}
+			
+			cJSON * json_address = cJSON_GetObjectItem(param, "address");
+			if (cJSON_IsString(json_address)) {
+				strcpy(address, json_address->valuestring);
+			}
+		} else {
+			ctx->error_code = 1;
+			ctx->error_message = strdup("Invalid parameters.");
+			return NULL;
+		}
+	} else {
+		ctx->error_code = 1;
+		ctx->error_message = strdup("Invalid parameters.");
+		return NULL;
+	}
+	
+	if(g_xdag_state < XDAG_STATE_XFER) {
```

Comment:
I really feel like we need OOP :) Most of this logic is similar to command.c. Something like base - derived classes.

Reply:
You can do OOP on C too :D